src/JustSaying/Messaging/Middleware/DefaultSqsMiddleware.cs

```diff
@@ -0,0 +1,63 @@
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Amazon.SQS.Model;
+using JustSaying.Messaging.Channels;
+using Microsoft.Extensions.Logging;
+
+namespace JustSaying.Messaging.Middleware
+{
+    public class DefaultSqsMiddleware : MiddlewareBase<GetMessagesContext, IList<Amazon.SQS.Model.Message>>
+    {
+        private readonly ILogger<DefaultSqsMiddleware> _logger;
+
+        public DefaultSqsMiddleware(ILogger<DefaultSqsMiddleware> logger) : base(null)
+        {
+            _logger = logger;
+        }
+
+        protected override async Task<IList<Message>> RunInnerAsync(GetMessagesContext context, Func<Task<IList<Message>>> func)
+        {
+            try
+            {
+                var results = await func().ConfigureAwait(false);
+
+                _logger.LogTrace(
+                    "Polled for messages on queue '{QueueName}' in region '{Region}', and received {MessageCount} messages.",
+                    context.QueueName,
+                    context.RegionName,
+                    context.Count);
+
+                return results;
+            }
+            catch (InvalidOperationException ex)
+            {
+                _logger.LogTrace(
+                    ex,
+                    "Could not determine number of messages to read from queue '{QueueName}' in '{Region}'.",
+                    context.QueueName,
+                    context.RegionName);
+            }
+            catch (OperationCanceledException ex)
+            {
+                _logger.LogTrace(
```

Comment:
Maybe this should be a warning?

Reply:
All of this logic is copy-pasta'd from [here](https://github.com/justeat/JustSaying/blob/25b5d8dbe3697b227168242abd2ada82a57bd88a/src/JustSaying/AwsTools/MessageHandling/SqsNotificationListener.cs#L148)

Reply:
We might need to review what errors we are actually throwing now and whether these are still expected in these scenarios

Reply:
`OperationCanceledException`'s thrown here will always be caused by a receive timeout from SQS, as its only handled in the context of the middleware.